chains/substrate/listener/listener.go

```diff
@@ -4,123 +4,30 @@
-	GetBlockEvents(hash types.Hash) ([]*parser.Event, error)
+type Connection interface {
 	GetFinalizedHead() (types.Hash, error)
 	GetBlock(blockHash types.Hash) (*types.SignedBlock, error)
+	GetBlockHash(blockNumber uint64) (types.Hash, error)
+	GetBlockEvents(hash types.Hash) ([]*parser.Event, error)
+	UpdateMetatdata() error
 }
 
-type SubstrateListener struct {
-	conn ChainConnection
-
-	eventHandlers      []EventHandler
-	blockRetryInterval time.Duration
-	blockInterval      *big.Int
-
-	log zerolog.Logger
-}
-
-func NewSubstrateListener(connection ChainConnection, eventHandlers []EventHandler, config *substrate.SubstrateConfig) *SubstrateListener {
-	return &SubstrateListener{
-		log:                log.With().Uint8("domainID", *config.GeneralChainConfig.Id).Logger(),
-		conn:               connection,
-		eventHandlers:      eventHandlers,
-		blockRetryInterval: config.BlockRetryInterval,
-		blockInterval:      config.BlockInterval,
-	}
-}
-
-func (l *SubstrateListener) ListenToEvents(ctx context.Context, startBlock *big.Int, domainID uint8, blockstore store.BlockStore, msgChan chan []*message.Message) {
-	endBlock := big.NewInt(0)
-
-	go func() {
-	loop:
-		for {
-			select {
-			case <-ctx.Done():
-				return
-			default:
-				hash, err := l.conn.GetFinalizedHead()
-				if err != nil {
-					l.log.Warn().Err(err).Msg("Failed to fetch finalized header")
-					time.Sleep(l.blockRetryInterval)
-					continue
-				}
-				head, err := l.conn.GetBlock(hash)
-				if err != nil {
-					l.log.Warn().Err(err).Msg("Failed to fetch block")
-					time.Sleep(l.blockRetryInterval)
-					continue
-				}
-
-				if startBlock == nil {
-					startBlock = big.NewInt(int64(head.Block.Header.Number))
-				}
-				endBlock.Add(startBlock, l.blockInterval)
-
-				// Sleep if finalized is less then current block
-				if big.NewInt(int64(head.Block.Header.Number)).Cmp(endBlock) == -1 {
-					time.Sleep(l.blockRetryInterval)
-					continue
-				}
-
-				evts, err := l.fetchEvents(startBlock, endBlock)
-				if err != nil {
-					l.log.Warn().Err(err).Msgf("Failed fetching events for block range %s-%s", startBlock, endBlock)
-					time.Sleep(l.blockRetryInterval)
-					continue
-				}
-
-				for _, handler := range l.eventHandlers {
-					err := handler.HandleEvents(evts, msgChan)
-					if err != nil {
-						l.log.Warn().Err(err).Msg("Error handling substrate events")
-						continue loop
-					}
-				}
-				err = blockstore.StoreBlock(endBlock, domainID)
-				if err != nil {
-					l.log.Error().Str("block", startBlock.String()).Err(err).Msg("Failed to write latest block to blockstore")
-				}
-				startBlock.Add(startBlock, l.blockInterval)
-			}
-		}
-	}()
-}
-
-func (l *SubstrateListener) fetchEvents(startBlock *big.Int, endBlock *big.Int) ([]*parser.Event, error) {
-	l.log.Debug().Msgf("Fetching substrate events for block range %s-%s", startBlock, endBlock)
+func FetchEvents(startBlock *big.Int, endBlock *big.Int, conn Connection) ([]*parser.Event, error) {
```

Comment:
Probably best to rename this file to something else. 
`listener.go` file names I would reserve  for things that actually listen. 
This is just an util function to fetch events.


Actually, it could be argued this should be in `sygma-core` as all implementations would need this. 
As part of the client for example.

Reply:
Needs some though for where it would make sense.